stage_descriptions/geospatial-04-cr3.md

```diff
@@ -0,0 +1,197 @@
+
+```
+LATITUDE_MIN = -85.05112878
+LATITUDE_MAX = 85.05112878
+LONGITUDE_MIN = -180
+LONGITUDE_MAX = 180
+```
+
+You can see how the official Redis implementation does this [here](https://github.com/redis/redis/blob/ff2f0b092c24d5cc590ff1eb596fc0865e0fb721/src/geohash.c#L141)
+
+2. The normalized values are then scaled in the range of [0, 2^26). This can be achieved by multiplying the normalized values by 2^26.
+
+```
+scaled_latitude = normalized_latitude * (2^26)
+scaled_longitude = normalized_longitude * (2^26)
+```
+
+You can see how the official Redis implementation does this [here](https://github.com/redis/redis/blob/ff2f0b092c24d5cc590ff1eb596fc0865e0fb721/src/geohash.c#L147)
+
+#### Interleaving bits
+
+1. The scaled values of latitude and longitude are converted to unsigned 32 bit unsigned integers. This is equivalent to dropping off the digits after the decimal point so that these values are now integers instead of floating point numbers.
+
+In the official Redis implementation, the scaled values of latitude and longitude (stored as `double`) are casted implicitly to `uint32_t` while calling `interleave64(uint32_t, uint32_t)`. You can view how this is done [here](https://github.com/redis/redis/blob/ff2f0b092c24d5cc590ff1eb596fc0865e0fb721/src/geohash.c#L149).
+
+
+2. The bits of latitude and longitude are interleaved to form a 64 bit unsigned integer.
+
+Let's consider the bits of latitude are `X31 X30 X29 ... X2 X1 X0` (32 bit integer), and the bits of longitude are `Y31 Y30 Y29 ... Y2 Y1 Y0` (32 bit integer). The end goal here is to construct a 64 bit integer whose digits will look like this:
+
+```
+score = Y31 X31 Y30 X30 Y29 X29 ... Y2 X2 Y1 X1 Y0 X0
+```
+
+The pseudocode below shows how this is done.
+
+```
+function interleave_bits(lat: uint32, lon: uint32) -> uint64:
+    # Cast the values to 64 bits unsigned integers
+    lat = uint_64(lat)
+    lon = uint_64(lon)
+
+    # Interleave-16
+    lat = (lat | (lat << 16)) & 0x0000FFFF0000FFFF
+    lon = (lon | (lon << 16)) & 0x0000FFFF0000FFFF
+
+    # Interleave-8
+    lat = (lat | (lat << 8))  & 0x00FF00FF00FF00FF
+    lon = (lon | (lon << 8))  & 0x00FF00FF00FF00FF
+
+    # Interleave-4
+    lat = (lat | (lat << 4))  & 0x0F0F0F0F0F0F0F0F
+    lon = (lon | (lon << 4))  & 0x0F0F0F0F0F0F0F0F
+
+    # Interleave-2
+    lat = (lat | (lat << 2))  & 0x3333333333333333
+    lon = (lon | (lon << 2))  & 0x3333333333333333
+
+    # Interleave-1
+    lat = (lat | (lat << 1))  & 0x5555555555555555
+    lon = (lon | (lon << 1))  & 0x5555555555555555
+
+    # Interleave latitude and lonngitude: latitude in even bits, longitude in odd bits
+    return lat | (lon << 1)
+```
+
+Let's focus only on the latitude value for now and call it x. This is a 32-bit integer made up of bits like X31, X30, ..., X0. The end goal here is to spread these bits out so that each one is separated by a zero. In other words, the goal here is to turn x into a 64-bit value: 0 X31 0 X30 0 X29 ... 0 X1 0 X0.
```

Comment:
Let's move this explanation into comments on the code block - explain just-in-time, avoid confusing users first and then explaining later